### PR TITLE
Add solution to Assignment 6

### PR DESCRIPTION
**Pull request by elifguler96**

### CODE REVIEW

File: Assignment6.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+public class Assignment6 {
+    /**
+     * Rearranges given to make it the same as desired.
+     *
+     * @param given   initial version of the array of numbers.
+     * @param desired final version of the array of numbers.
+     * @return list of moves made during arrangement.
+     */
+    public static List<List<Integer>> rearrange(int[] given, int[] desired) {
```
Comment (fulopbbea) [inline Assignment6.java]: The return value should be something more precise e.g List<Move> where Move is a simple data structure containing info on a move (car no., slot no. and empty slot no. for instance)

File: Assignment6.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        for (int i = 0; i < given.length; i++) {
+            indexMap.put(given[i], i);
+        }
+
+        for (int i = 0; i < desired.length; i++) {
+            // Already in its desired place.
+            if (indexMap.get(desired[i]) == i) {
+                continue;
+            }
+
```
Comment (fulopbbea) [inline Assignment6.java]: I guess we should test it if the current slot (i) should be the empty slot. In this case there should be no moves to be added. eg. {1, 2, 0, 3}->{0, 1, 3, 2} would generate something like car no. 0 moves which is not allowed

File: Assignment6.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            int indexOf0 = indexMap.get(0);
+
+            // If i isn't already empty.
+            if (indexOf0 != i) {
+                // Move the number in i to the empty slot so that i becomes empty.
+                indexMap.put(0, i);
```
Comment (fulopbbea) [inline Assignment6.java]: The code would be more readable and less error prone if we had a helper method called move for instance which performs a move both on the map and the array.

File: Assignment6.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                indexMap.put(given[i], indexOf0);
+
+                given[indexOf0] = given[i];
+                given[i] = 0;
+
+                printMove(i, indexOf0);
```
Comment (fulopbbea) [inline Assignment6.java]: Maybe we should have a separate function which prints the rearrangement plan separately in order to have methods that have only one clearly defined task each.

File: Assignment6Test.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
```
Comment (fulopbbea) [inline Assignment6Test.java]: Maybe we should add some more test cases to make sure we cover all the code. Maybe we should also add some longer examples.

File: Assignment6.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+     * @param desired final version of the array of numbers.
+     * @return list of moves made during arrangement.
+     */
+    public static List<List<Integer>> rearrange(int[] given, int[] desired) {
+        // Keys are values in given, values are their indexes in given.
+        Map<Integer, Integer> indexMap = new HashMap<>();
+        // Generated for testing purposes.
+        List<List<Integer>> listOfMoves = new LinkedList<>();
+
+        for (int i = 0; i < given.length; i++) {
```
Comment (fulopbbea) [inline Assignment6.java]: Populating the map could also go into a helper method for readability purposes. :)

**Review by fulopbbea — CHANGES_REQUESTED** [review]
I added the comments we discussed about earlier. I guess you could also use snippets/ideas from the other solutions we coded to get it done faster. :) Good job! :D

File: Assignment6.java (hunk 4 above)
Comment (elifguler96) [reply Assignment6.java]: The `printMove` method's only task is to print the move. I didn't really understand what you wanted to say.

Comment (fulopbbea) [reply Assignment6.java]: I meant that maybe we shouldn't print anything at all in the rearrange function but to write a separate function that prints the full list of moves afterwards. That would make more sense to me, but should be ok either way. :)

**Comment by elifguler96** [comment]
Another thought: we could just get the index of the empty slot and put the car that is supposed to be there to that index. It should minimize the number of moves.

The only place this would be stuck is if the empty slot is in its correct place (ex. given: 2 0 3 1 desired: 2 0 1 3). To solve this, we could have a queue which has all the cars that are not in its correct place (1 and 3 in this example) and move the first car in the queue to the empty slot. And when we change the place of the empty slot, we can just continue on doing what we were doing. 
We can also put the car we polled back in the queue but I'm not really sure whether it is necessary.

We should also have an array or something to know whether the car is in its correct place or not. There could be a boolean array for this.

To be clear, we don't have to search through the queue and remove a car it is moved to its correct place. The queue is only to be used in the empty slot being in its correct place case. When we poll a car from the queue, we can check if it is in its correct place (from the flag array mentioned above), if it is, we just poll another car. (We don't put that car back in the queue)

I think both time and space complexity of this would be O(N)

PS. check out the slack channel if you missed the discussion 😄

**Comment by cba33** [comment]
One idea to implement car swap using a minimum nr of moves (following our discussion on slack):
1. we first do the greedy approach suggested by @elifguler96  and @fulopbbea to get cars on their spot in one move each. we can't do better than one move per misplaced car. we perform this step until '0' is on its correct spot in the output array.
2. after that, when 0 is on it's final spot, we simply take the remaining misplaced cars in pairs in the order given by the output array - getting a pair of cars on their spots would take 3 moves, and that's optimal because we cannot do it it two moves for a pair of cars, otherwise we would have done it in the first step. we perform this step for a single pair of cars. Continue with step 1.

Why it's optimal: we move misplaced cars on their final spot either in one move (step 1) or 1.5 moves per misplaced car (step 2). We only perform step 2 when we cannot perform step 1.

Running demo: given 1 2 3 4 5 6 0 as input and 3 1 6 4 2 5 0 as output
- We cannot apply rule 1, so we apply rule 2 for pair <3,1>: 
           1 2 3 4 5 6 0 -> 1 0 3 4 5 6 2 -> 0 1 3 4 5 6 2 -> 3 1 0 4 5 6 2
- Apply rule 1 and get 6 on its spot: 3 1 0 4 5 6 2 -> 3 1 6 4 5 0 2
- Apply rule 1 and get 5 on its spot: 3 1 6 4 5 0 2 -> 3 1 6 4 0 5 2
- Apply rule 1 and get 2 on its spot: 3 1 6 4 0 5 2 -> 3 1 6 4 2 5 0
- All cars are on their spot, stop

**Comment by elifguler96** [comment]
I really like your idea @cba33 it is far more better than just picking a randomly and putting it in the empty slot

**Comment by fulopbbea** [comment]
@cba33 I'm not really sure about the 2nd rule, I'd say it could be more than a pair which gets reduced to the first step except when we have to unblock the algorithm because of the empty slot

**Comment by katermar** [comment]
@elifguler96 I really like your suggestions, but maybe we can do it without a queue and extra space for it just finding first mismathing element which is not an empty lot? Then? as you've said, we can swap this misplaced element into empty position. In the next iteration after this hotfix is guaranteed that you change an element in the right place. Thus, even if the particular case is, it still only takes 2 swap to move element out of place.

**Comment by elifguler96** [comment]
@katermar but that would make the algorithm O(N^2). I added the queue to make it O(N)

**Comment by cba33** [comment]
@katermar i think your idea works

**Comment by adifeiner** [comment]
Another thought, we can use BiMap (HashBiMap) for the given and desired array. it can help us know when we finish and to find the next Car in a readable way. When the given is the same like inverse desired we know we've finished. 
Explanation about BiMap (Google guava library): https://github.com/google/guava/wiki/NewCollectionTypesExplained#bimap

**Pull request merged**